### PR TITLE
[Enhancement] Support bitmap_hash for count(distinct x) rewrite

### PR DESCRIPTION
**Pull request by LiShuMing**
Why I'm doing:
- Only support bitmap_union(to_bitmap(x)) rewrite for count(distinct x) which x must be numeric column or string column that can be converted to numeric
- If column x is string column, we cannot rewrite bitmap_union(bitmap_hash(x)) for count distinct

What I'm doing:
- Support bitmap_hash for count(distinct x) rewrite
- Refactor `AggregateFunctionRewriter` into `IRewriteEquivalent` to support more cases.
```
-- MV
select user_id, bitmap_union(bitmap_hash(user_name)) from user_tags group by user_id;
-- Query
select user_id, count(distinct user_name) x from user_tags group by user_id;
```
## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/equivalent/CountRewriteEquivalent.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rule.transformation.materialization.AggregatedMaterializedViewRewriter;
+
+public class CountRewriteEquivalent extends IAggregateRewriteEquivalent {
+    public static IRewriteEquivalent INSTANCE = new CountRewriteEquivalent();
+
+    public CountRewriteEquivalent() {}
+
+    public boolean check(ScalarOperator op) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/equivalent/CountRewriteEquivalent.java:28]:
```suggestion
    private boolean check(ScalarOperator op) {
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/equivalent/IPredicateRewriteEquivalent.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ *   b. dt >= '2023-11-21 00:00:00'
+ * example2:
+ *   a. date_trunc('hour', dt) >= '2023-11-21 00:00:00'
+ *   b. dt >= '2023-11-21 00:00:00'
+ *
+ * {@code IRewriteEquivalent} is used to check whether different expression are equivalent or not.
+ */
+public abstract class IPredicateRewriteEquivalent implements IRewriteEquivalent {
+    public RewriteEquivalentType getRewriteEquivalentType() {
+        return RewriteEquivalentType.PREDICATE;
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/equivalent/IPredicateRewriteEquivalent.java:33]: what is this api for?

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/equivalent/IPredicateRewriteEquivalent.java:33]: Add more docs about it:
```
    /**
     * Different {@code RewriteEquivalentType} will be used to rewrite different scalar operators.
     * eg:
     *  `PREDICATE` will be used to rewrite for `BinaryPredicateOperator`;
     *  `CallOperator` will be used to rewrite for `BinaryPredicateOperator`;
     */
```

**Review by ABingHuang — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by LiShuMing** [comment]
@mergifyio backport branch-3.1

**Comment by LiShuMing** [comment]
@mergifyio backport to branch-3.1

**Comment by LiShuMing** [comment]
@mergifyio backport branch-3.1

**Pull request merged**